File: src/main/java/tmi/recipe/parser/FrackerParser.java
package tmi.recipe.parser;

import arc.math.Mathf;
import arc.struct.Seq;
import mindustry.Vars;
import mindustry.world.Block;
import mindustry.world.blocks.environment.Floor;
import mindustry.world.blocks.production.Fracker;
import tmi.recipe.Recipe;
import tmi.recipe.RecipeType;

public class FrackerParser extends ConsumerParser<Fracker>{
  {
    excludes.add(PumpParser.class);
    excludes.add(SolidPumpParser.class);
  }

  @Override
  public boolean isTarget(Block content) {
    return content instanceof Fracker;
  }

  @Override
  public Seq<Recipe> parse(Fracker fracker) {
    Recipe res = new Recipe(RecipeType.collecting)
        .setEfficiency(Recipe.getDefaultEff(fracker.baseEfficiency))
        .setBlock(getWrap(fracker))
        .setTime(fracker.consumeTime);

    res.addProductionPresec(getWrap(fracker.result), fracker.pumpAmount);

    registerCons(res, fracker.consumers);

    for (Block block : Vars.content.blocks()) {
      if (block.attributes.get(fracker.attribute) <= 0 || (block instanceof Floor f && f.isDeep())) continue;

      float eff = block.attributes.get(fracker.attribute);
      res.addMaterialRaw(getWrap(block), fracker.size*fracker.size)
          .setOptionalCons(fracker.baseEfficiency > 0.001f)
          .setEfficiency(eff)
          .setAttribute()
          .setFormat(f -> "[#98ffa9]" + (fracker.baseEfficiency > 0.001f? "+": "") + Mathf.round(eff*100) + "%");
    }

    return Seq.with(res);
  }
}

File: src/main/kotlin/tmi/recipe/parser/FrackerParser.kt
package tmi.recipe.parser

import arc.struct.Seq
import mindustry.Vars
import mindustry.world.Block
import mindustry.world.blocks.environment.Floor
import mindustry.world.blocks.production.Fracker
import tmi.recipe.Recipe
import tmi.recipe.RecipeItemGroup
import tmi.recipe.types.RecipeItemType
import tmi.recipe.RecipeType

open class FrackerParser : ConsumerParser<Fracker>() {
  init {
    excludes.add(PumpParser::class.java)
    excludes.add(SolidPumpParser::class.java)
  }

  override fun isTarget(content: Block): Boolean {
    return content is Fracker
  }

  override fun parse(content: Fracker): Seq<Recipe> {
    val res = Recipe(
      recipeType = RecipeType.collecting,
      ownerBlock = content.getWrap(),
      craftTime = content.itemUseTime,
    ).setBaseEff(content.baseEfficiency)

    res.addProductionPersec(content.result.getWrap(), content.pumpAmount)

    registerCons(res, *content.consumers)

    val attrGroup = RecipeItemGroup()
    for (block in Vars.content.blocks()) {
      if (content.attribute == null || block.attributes[content.attribute] <= 0 || (block is Floor && block.isDeep)) continue

      val eff = block.attributes[content.attribute]
      res.addMaterial(block.getWrap(), (content.size*content.size) as Number)
        .setOptional(content.baseEfficiency > 0.001f)
        .setEff(eff)
        .setType(RecipeItemType.ATTRIBUTE)
        .efficiencyFormat(content.baseEfficiency, eff)
        .setGroup(attrGroup)
    }

    return Seq.with(res)
  }
}
